Updating Native Modules Documentation

lauramedalia opened the pull request:
To Do
- Make Note wording consistent  (Note: vs NOTE etc)
- Constituency around what is code formatted or not (file names or data types for example)
- Move synchronous methods to “beyond a calendar module” section?
- FYI seperated side bar into 2 parts
  - <img width="275" alt="Screen Shot 2020-10-19 at 11 47 48 PM" src="https://user-images.githubusercontent.com/37278825/96538104-c28dc580-1265-11eb-80ec-482540f73a4c.png">
- Callback order

charpeni commented:
Deploy preview for *react-native* ready!

Built with commit e87ce68219e6d9c466b6eeb0d14b5c82ebe6e701

https://deploy-preview-2261--react-native.netlify.app

lauramedalia commented on website/versioned_docs/version-0.63/native-modules-ios.md:
```diff
@@ -1,443 +1,467 @@
 ---
 
-Sometimes an app needs to access a platform API and React Native doesn't have a corresponding module yet. Maybe you want to reuse some existing Objective-C, Swift or C++ code without having to reimplement it in JavaScript, or write some high performance, multi-threaded code such as for image processing, a database, or any number of advanced extensions.
+If you haven’t already, please start by reading the [Native Modules Intro](native-modules-intro).
```
Is. the following a bit more welcoming: "Welcome to Native Modules for iOS. Please start by reading the [Native Modules Intro](native-modules-intro) for an intro to what native modules are."

RSNara commented on docs/native-modules-android.md:
```diff
@@ -3,113 +3,124 @@ id: native-modules-android
+
+### Module Name
+
+All native modules in Android need to implement the `getName()` method. This method returns the string name of the native module which represents the class in JavaScript. The native module can then be accessed in JS via its name. For example, in the below code snippet, `getName` returns `"CalendarModule"`.
```
What's the convention on writing method names: Do we always suffix with `()`?

lauramedalia replied:
I am not sure, @rachelnabors  is there an opinion here? I removed () from all method names in the docs to be consistent for now!

rachelnabors replied:
We use `()` to help distinguish methods and functions from objects at a glance. (Unless you can think of a reason not to!)

lauramedalia replied:
just updated to include them

RSNara commented on docs/native-modules-ios.md:
````diff
@@ -3,440 +3,464 @@ id: native-modules-ios
 ```
 
-In addition to implementing the `RCTBridgeModule` protocol, your class must also include the `RCT_EXPORT_MODULE()` macro. This takes an optional argument that specifies the name that the module will be accessible as in your JavaScript code (more on this later). If you do not specify a name, the JavaScript module name will match the Objective-C class name. If the Objective-C class name begins with ...
+You can use any name that fits the native module you are building. Name the interface `RCTCalendarModule` since you are creating a calendar native module. Since ObjC does not have namespaces like Java or C++, convention is to prepend the interface name with a substring like RCT (an abbreviation or React).
````
Might be a good idea to indicate that the substring is application-specific.

RSNara commented on docs/native-modules-ios.md:
```diff
@@ -464,36 +488,91 @@ You will receive a warning if you expend resources unnecessarily by emitting an
+
+### Exporting Swift
+
+Swift doesn't have support for macros so exposing it to React Native requires a bit more setup but works relatively the same. Let's say we have the same `CalendarModule` but as a Swift class:
```
```suggestion
Swift doesn't have support for macros, so exposing native modules and their methods to JavaScript inside React Native requires a bit more setup. However, it works relatively the same. Let's say we have the same `CalendarModule` but as a Swift class:
```

RSNara commented on docs/native-modules-ios.md:
```diff
@@ -515,10 +594,10 @@ For those of you new to Swift and Objective-C, whenever you [mix the two languag
+#### invalidate()
 
-Native modules can conform to the [RCTInvalidating](https://github.com/facebook/react-native/blob/master/React/Base/RCTInvalidating.h) protocol on iOS by implementing the `invalidate` method. This method can be invoked when the native bridge is invalidated (ie: on devmode reload). You should avoid implementing this method in general, as this mechanism exists for backwards compatibility and may be ...
+Native modules can conform to the [RCTInvalidating](https://github.com/facebook/react-native/blob/0.62-stable/React/Base/RCTInvalidating.h) protocol on iOS by implementing the `invalidate` method. This method [can be invoked](https://github.com/facebook/react-native/blob/0.62-stable/ReactCommon/turbomodule/core/platform/ios/RCTTurboModuleManager.mm#L456) when the native bridge is invalidated (ie: ...
```
```suggestion
Native modules can conform to the [RCTInvalidating](https://github.com/facebook/react-native/blob/0.62-stable/React/Base/RCTInvalidating.h) protocol on iOS by implementing the `invalidate` method. This method [can be invoked](https://github.com/facebook/react-native/blob/0.62-stable/ReactCommon/turbomodule/core/platform/ios/RCTTurboModuleManager.mm#L456) when the native bridge is invalidated (ie: on devmode reload). Please use this mechanism as necessary to do the required cleanup for your native module. 
```

RSNara reviewed (commented):
Reviewed the iOS docs. Just a few minor suggestions! No correctness mistakes. Will take a look at Android soon.

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
```diff
@@ -1,116 +1,127 @@
 
-  private static final String DURATION_SHORT_KEY = "SHORT";
-  private static final String DURATION_LONG_KEY = "LONG";
+> Note: It is worth noting that technically Java classes can extend the `BaseJavaModule` class or implement the `NativeModule` interface to be considered a Native Module by React Native.
```
```suggestion
> Note: It is worth noting that technically Java classes only needs to extend the `BaseJavaModule` class or implement the `NativeModule` interface to be considered a Native Module by React Native.
```

lauramedalia replied:
Applying this in a separate commit to fix a grammar issue! (should be: only need to)

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
```diff
@@ -120,236 +131,420 @@ import java.util.ArrayList;
-The package needs to be provided in the `getPackages` method of the `MainApplication.java` file. This file exists under the android folder in your react-native application directory. The path to this file is: `android/app/src/main/java/com/your-app-name/MainApplication.java`.
+This file imports the native module we created, `CalendarModule`. It then instantiates `CalendarModule` within the `createNativeModules` function and returns it as a list of `NativeModules` to register. If you add more native modules down the line, you can also instantiate them and add them to the list returned here.
+
+> Note: It is worth noting that this way of registering native modules eagerly initializes all native modules when the application starts, which adds to the startup time of an application. You can use [TurboReactPackage](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/TurboReactPackage.java) as an alternative. Instead of `createNativeModules`, wh ...
```
Might be a good idea to just use the full GitHub links, as opposed to redirects.

```suggestion
> Note: It is worth noting that this way of registering native modules eagerly initializes all native modules when the application starts, which adds to the startup time of an application. You can use [TurboReactPackage](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/TurboReactPackage.java) as an alternative. Instead of `createNativeModules`, which return a list of instantiated native module objects, TurboReactPackage implements a `getModule(String name, ReactApplicationContext rac)` method that creates the native module object, when required. TurboReactPackage is a bit more complicated to implement at the moment. In addition to implementing a `getModule` method, you have to implement a `getReactModuleInfoProvider()` method, which returns a list of all the native modules the package can instantiate along with a function that instantiates them, example [here](https://github.com/facebook/react-native/blob/8ac467c51b94c82d81930b4802b2978c85539925/ReactAndroid/src/main/java/com/facebook/react/CoreModulesPackage.java#L86-L165). Again, using TurboReactPackage will allow your application to have a faster startup time, but it is currently a bit cumbersome to write. So proceed with caution if you choose to use TurboReactPackages.
```

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
```diff
@@ -445,29 +640,33 @@ public class ImagePickerModule extends ReactContextBaseJavaModule {
+
+To date, on Android, all native module async methods execute on one thread. On iOS, each native module can provide its method queue, or have React Native generate a method queue for it. On iOS, native modules also have an API to get a hold of the method queue on which their async methods are executed.
+
+Native modules should not have any assumptions about what thread they are being called on, as the current assignment is subject to change in the future. If a blocking call is required, the heavy work should be dispatched to an internally managed worker thread, and any callbacks distributed from there.
```
It looks like we don't mention the `invalidate` Android analogue. In Java, NativeModules can have a `onCatalystInstanceDestroy()` method that they can implement to do cleanup of any resources they own. There's also an initialize method: https://github.com/facebook/react-native/blob/ae418b83c071d27d1a5bc7dcb46d546686bd21b8/ReactAndroid/src/main/java/com/facebook/react/bridge/NativeModule.java#L40-L51.

lauramedalia replied:
@RSNara I ended up removing the notes about iOS here. Do you want to add this info about Android here?

RSNara reviewed (commented):
Looks good! Left a few suggestions.

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
````diff
@@ -1,116 +1,127 @@
+}
+```
+
+As you can see, your `CalendarModule` class extends the `ReactContextBaseJavaModule`. For Android, most native modules consist of a class that extends `ReactContextBaseJavaModule` and implements the functionality required by Javascript.
````
The word "consist", how it's used here, is a bit confusing: "For Android, most native modules consist of a class". 

```suggestion
As you can see, your `CalendarModule` class extends the `ReactContextBaseJavaModule` class. For Android, Java native modules are written as classes that extend `ReactContextBaseJavaModule` and implement the functionality required by Javascript.
```

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
```diff
@@ -1,116 +1,127 @@
+
+### Module Name
+
+All native modules in Android need to implement the `getName()` method. This method returns the string name of the native module which represents the class in JavaScript. The native module can then be accessed in JS via its name. For example, in the below code snippet, `getName` returns `"CalendarModule"`.
```
```suggestion
All Java native modules in Android need to implement the `getName()` method. This method returns a string, which represents the name of the native module. The native module can then be accessed in JavaScript using its name. For example, in the below code snippet, `getName()` returns `"CalendarModule"`.
```

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
````diff
@@ -1,116 +1,127 @@
 ```
 
-### Argument Types
+Once we finish implementing the native module and hook it up in JS, you can follow [these steps](https://developer.android.com/studio/debug/am-logcat.html) to view the logs from your app.
````
```suggestion
Once we finish implementing the native module and hook it up in JavaScript, you can follow [these steps](https://developer.android.com/studio/debug/am-logcat.html) to view the logs from your app.
```

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
```diff
@@ -1,116 +1,127 @@
+### Register the Module (Android Specific)
 
-The last step within Java is to register the Module; this happens in the `createNativeModules` of your apps package. If a module is not registered it will not be available from JavaScript.
+Once a native module is written, it needs to be registered with React Native. In order to do so, you need to add your native module to a `ReactPackage` and register the `ReactPackage` with React Native. During initialization, React Native will loop over all packages, and for each `ReactPackage`, register each native module within.
```
> register each native module within.

After reading this, I'm left wondering what it means to register a NativeModule with React Native. Maybe we should elaborate on this. What we're really doing is making each NativeModule available to JavaScript code.

**Edit:** Looks like we already talk about this down the line.

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
```diff
@@ -120,236 +131,420 @@ import java.util.ArrayList;
-}
+### Test What You Have Built
+
+At this point, you have set up the basic scaffolding for your native module in Android. Test that out by accessing the native module and invoking it’s exported method in JavaScript.
```
```suggestion
At this point, you have set up the basic scaffolding for your native module in Android. Test that out by accessing the native module and invoking its exported method in JavaScript.
```

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
````diff
@@ -120,236 +131,420 @@ import java.util.ArrayList;
 ```
 
-To access your new functionality from JavaScript, it is common to wrap the native module in a JavaScript module. This is not necessary but saves the consumers of your library the need to pull it off of `NativeModules` each time. This JavaScript file also becomes a good location for you to add any JavaScript side functionality.
+In order to access your native module from Javascript you need to first import `NativeNodules` from React Native:
````
```suggestion
In order to access your native module from Javascript you need to first import `NativeModules` from React Native:
```

lauramedalia replied:
Nodules!! Gah good catch 🤪

RSNara commented on website/versioned_docs/version-0.63/native-modules-android.md:
```diff
@@ -120,236 +131,420 @@ import java.util.ArrayList;
 ### Callbacks
 
-Native modules also support a unique kind of argument - a callback. In most cases it is used to provide the function call result to JavaScript.
+Native modules also support a unique kind of argument: a callback. Callbacks are used to pass data from Java to Javascript for asynchronous methods. They can also be used to asynchronously execute JS from the native side.
```
```suggestion
Native modules also support a unique kind of argument: a callback. Callbacks are used to pass data from Java to JavaScript for asynchronous methods. They can also be used to asynchronously execute JavaScript from the native side.
```

lauramedalia commented on docs/native-modules-ios.md:
```diff
@@ -3,440 +3,464 @@ id: native-modules-ios
-> **NOTE**: JavaScript method names
->
-> The name of the method exported to JavaScript is the native method's name up to the first colon. React Native also defines a macro called `RCT_REMAP_METHOD()` to specify the JavaScript method's name. This is useful when multiple native methods are the same up to the first colon and would have conflicting JavaScript names.
+If you do not specify a name, the JavaScript module name will match the Objective-C class name. Please note, if the Objective-C class name begins with RCT, the JavaScript module name will exclude the RCT prefix.
```
Like the suggestion but yes I don't have context on the RK prefix

lauramedalia commented on docs/native-modules-ios.md:
````diff
@@ -3,440 +3,464 @@ id: native-modules-ios
 ```
 
-or like this:
+> Please note that the `RCT_EXPORT_METHOD` macro will not be necessary with TurboModules unless your method relies on RCT argument conversion (see argument types below). Ultimately we would like to get rid of the `RCT_EXPORT_MACRO` and therefore discourage people from using `RCTConvert`. Instead, you can do the argument conversion within the method body.
````
Yes! I love this idea. We had chatted about adding a section at the beginning as well that noted what turbo modules is. I am happy with either/or but would love a partner to come up with the summary. Maybe something like the following as a note at the beginning of the Android/iOS docs?

The React Native team is currently working on a re-architecture of React Native called TurboModules. TurboModules will help React Native do X,Y,Z, you can read more about it here. Throughout these docs we have added notes around parts of Native Modules that will change in the TurboModules release and how you can best prepare for a smooth upgrade to TurboModules.

lauramedalia replied:
@RSNara  @rachelnabors  I think this got lost in the PR, what are your thoughts here? I can add this note on the top

lauramedalia commented on docs/native-modules-ios.md:
```diff
@@ -3,440 +3,464 @@ id: native-modules-ios
-  'Birthday Party',
-  '4 Privet Drive, Surrey'
-);
+const { CalendarModuleFoo } = ReactNative.NativeModules;
```
Just updated!

Simek commented:
Hi @lauramedalia, 

Yesterday we have finished the site migration process to the new version of Docusaurus and currently your PR is using the old website (V1) as a base.

Can you rebase your PR branch against `master` and resolve the conflicts? 🙂 This should make the future merge easier and it will allow you to test the changes in the new environment.

lauramedalia commented on docs/native-modules-ios.md:
```diff
@@ -3,440 +3,464 @@ id: native-modules-ios
-In addition to implementing the `RCTBridgeModule` protocol, your class must also include the `RCT_EXPORT_MODULE()` macro. This takes an optional argument that specifies the name that the module will be accessible as in your JavaScript code (more on this later). If you do not specify a name, the JavaScript module name will match the Objective-C class name. If the Objective-C class name begins with ...
+You can use any name that fits the native module you are building. Name the interface `RCTCalendarModule` since you are creating a calendar native module. Since ObjC does not have namespaces like Java or C++, convention is to prepend the interface name with a substring like RCT (an abbreviation or React).
+
+As you can see below, the CalendarModule interface implements the `RCTBridgeModule` protocol. A native module is an Objective-C class that implements the `RCTBridgeModule` protocol.
```
Forgot to respond, but yes nice call I updated this!

lauramedalia commented on docs/native-modules-ios.md:
````diff
@@ -464,36 +488,91 @@ You will receive a warning if you expend resources unnecessarily by emitting an
+
+React Native will initialize any registered `RCTBridgeModules` automatically, however you may wish to instantiate your own module instances (so you may inject dependencies, for example).
+
+You can do this by creating a class that implements the `RCTBridgeDelegate` Protocol, initializing an `RCTBridge` with the delegate as an argument and initialising a `RCTRootView` with the initialized bridge.
+
+```objectivec
+id<RCTBridgeDelegate> moduleInitialiser = [[classThatImplementsRCTBridgeDelegate alloc] init];
+
+RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:moduleInitialiser launchOptions:nil];
+
+RCTRootView *rootView = [[RCTRootView alloc]
+                        initWithBridge:bridge
+                            moduleName:kModuleName
+                     initialProperties:nil];
+```
````
@RSNaraj just to confirm - do you want to replace the first 2 paragraphs in this section with that text?

lauramedalia commented:
@RSNara  @rachelnabors  bringing this convo back up here

@RSNara  
This is awesome! :)

I wonder how many of these TurboModules tips we have. I also wonder if we should standardize on how we deliver them, so that they're easy to spot in the docs.

 @lauramedalia
lauramedalia 25 days ago Author
Yes! I love this idea. We had chatted about adding a section at the beginning as well that noted what turbo modules is. I am happy with either/or but would love a partner to come up with the summary. Maybe something like the following as a note at the beginning of the Android/iOS docs?

The React Native team is currently working on a re-architecture of React Native called TurboModules. TurboModules will help React Native do X,Y,Z, you can read more about it here. Throughout these docs we have added notes around parts of Native Modules that will change in the TurboModules release and how you can best prepare for a smooth upgrade to TurboModules.

Simek commented on docs/native-modules-ios.md:
```diff
@@ -1,442 +1,466 @@
+## Create a Calendar Native Module
 
-This is a more advanced guide that shows how to build a native module. It assumes the reader knows Objective-C or Swift and core libraries (Foundation, UIKit).
+In the following guide you will create a native module, `CalendarModule`, that will allow you to access Apple's calendar APIs from Javascript. By the end you will be able to call `CalendarModule.createCalendarEvent(‘Dinner Party’, ‘My House’);` from JavaScript, invoking a native method that creates a calendar event.
```
Apostrophe characters are incorrect in the code here (which will lead to syntax errors). 

The typography correct version has been used (`‘` and `’`) instead of simple `'`.

lauramedalia replied:
updated

Simek commented on docs/native-modules-ios.md:
```diff
@@ -1,442 +1,466 @@
+When a native module method is invoked in Javascript, React Native converts the arguments from JS objects to their Objective-C/Swift object analogues. So for example, if your Objective-C Native Module method accepts a NSNumber, in JS you need to call the method with a number. React Native will handle the conversion for you. Below is a list of the argument types supported for native module methods ...
 
-The native module should not have any assumptions about what thread it is being called on. React Native invokes native modules methods on a separate serial GCD queue, but this is an implementation detail and might change. The `- (dispatch_queue_t)methodQueue` method allows the native module to specify which queue its methods should be run on. For example, if it needs to use a main-thread-only iOS ...
+- string -> NSString \*
```
Can you convert this list into the table? I think this will improve the readability.

Also what's the purpose of asterisk at the end of some entries?

lauramedalia replied:
Updated to tables. Not sure where that asterisk came from, I had copied over from the original docs.

Simek replied:
It looks better, thank you! 

One more question, about for example `NSArray` - there are two very similar rows for this type, can this be simplified by putting in JS column `?Array | Array` or just `?Array`? (This hint might be wrong since I'm not that experienced with RN conversions)

<img width="607" alt="Screenshot 2020-11-28 145722" src="https://user-images.githubusercontent.com/719641/100517309-0ab7d600-318a-11eb-94f7-3c4c8c175763.png">

lauramedalia replied:
@Simek followed up with the table params here: https://github.com/facebook/react-native-website/pull/2373

Simek commented on docs/native-modules-ios.md:
```diff
@@ -1,442 +1,466 @@
-  time: date.getTime(),
-  description: '...'
-});
+react-native run-ios
```
We should not recommend using global CLI, please convert this line to `npx react-native run-ios` and change the code block language from `jsx` to `shell`.

lauramedalia replied:
done

Simek commented on docs/native-modules-ios.md:
```diff
@@ -1,442 +1,466 @@
+### Building as You Iterate
 
-## Callbacks
+As you work through these guides and iterate on your native module, you will need to do a native rebuild of your application to access your most recent changes from Javascript. This is because the code that you are writing sits within the native part of your application. While React Native’s metro bundler can watch for changes in Javascript and rebuild JS bundle on the fly for you, it will not do ...
```
Again, please change the run command to `npx react-native run-ios`.

lauramedalia replied:
done

Simek commented on docs/native-modules-android.md:
```diff
@@ -1,115 +1,128 @@
+## Create a Calendar Native Module
 
-## Native Module Setup
+In the following guide you will create a native module, `CalendarModule`, that will allow you to access Android’s calendar APIs from JavaScript. By the end, you will be able to call `CalendarModule.createCalendarEvent(‘Dinner Party’, ‘My House’);` from JavaScript, invoking a Java method that creates a calendar event.
```
The typography correct version of apostrophe has been used (`‘` and `’`) instead of simple one `'` (same issue like in iOS guide).

lauramedalia replied:
updated

Simek commented on docs/native-modules-android.md:
````diff
@@ -1,115 +1,128 @@
-import android.widget.Toast;
+Then add the following content:
 
+```jsx
````
Please change back the code block language from `jsx` to `java`.

lauramedalia replied:
done

Simek commented on docs/native-modules-android.md:
````diff
@@ -119,236 +132,420 @@ import java.util.ArrayList;
+Native modules can signal events to JavaScript without being invoked directly. For example, you might want to signal to Javascript a reminder that a calendar event from the native Android calendar app will occur soon. The easiest way to do this is to use the `RCTDeviceEventEmitter` which can be obtained from the `ReactContext` as in the code snippet below.
 
 ```java
+
````
Please clean up the unnecessary new line/whitespaces at the beginning and at the end of the code blocks.

lauramedalia replied:
done

Simek commented on docs/native-modules-android.md:
````diff
@@ -119,236 +132,420 @@ import java.util.ArrayList;
 
-Create a new JavaScript file named `ToastExample.js` with the content below:
+```jsx
+react-native run-android
````
Same like in iOS guide, please use the `npx` for the run command.

lauramedalia replied:
done

Simek commented on website/i18n/en.json:
```diff
@@ -0,0 +1,4325 @@
+{
```
This file no longer exist in V2 (i18n will be added in the future), so please remove it form this changeset.

lauramedalia replied:
done

Simek reviewed (changes requested):
Thank you for this contribution, that's a great improvement! 👏 

Please address the the minor issues from the comments above 🙂

Simek commented on website/versioned_docs/version-0.60/native-modules-intro.md:
```diff
@@ -0,0 +1,26 @@
+---
+id: version-0.5-native-modules-intro
```
`id` parameter is incorrect, also `original_id` parameter is redundant.

Also if you would like to update the versioned docs for `0.60` you also need to update the docs for all versions in between (`0.61`, `0.62`).

lauramedalia replied:
@Simek removing this, I am only adding native-modules-intro + the new docs it links to in 0.63

Simek commented on website/versioned_docs/version-0.63/native-modules-intro.md:
```diff
@@ -0,0 +1,26 @@
+---
+id: native-modules-intro
+title: Native Modules Intro
+original_id: native-modules-intro
```
`original_id` is no longer needed in V2 - https://v2.docusaurus.io/docs/migrating-from-v1-to-v2/#migrate-your-versioned_docs-front-matter

lauramedalia replied:
removed

lauramedalia commented:
> > Yes! I love this idea. We had chatted about adding a section at the beginning as well that noted what turbo modules is. I am happy with either/or but would love a partner to come up with the summary. Maybe something like the following as a note at the beginning of the Android/iOS docs?
> 
> The React Native team is currently working on a re-architecture of the NativeModule system. This new system is called TurboModules, and it will help facilitate more efficient type-safe communication between JavaScript and native, without relying on the React Native bridge. It will also enable new extensions that weren't possible with the legacy NativeModule system. You can read more about it here. Throughout these docs we have added notes around parts of Native Modules that will change in the TurboModules release and how you can best prepare for a smooth upgrade to TurboModules.

@RSNara where was the "here" supposed to link to? I see in a release doc you linked to here - is this what you intended? https://github.com/react-native-community/discussions-and-proposals/issues/40

lauramedalia commented:
> Thank you for this contribution, that's a great improvement! 👏
> 
> Please address the the minor issues from the comments above 🙂

@Simek addressed feedback, LMK if it looks good!

Simek commented:
@lauramedalia Thank you so much for the updates! 👍 

I have left one more question about the types table in non-collapsed comment above. Other than that this changeset LGTM!

Hope that other reviewers are also happy about the changes. 😉

rachelnabors reviewed (approved):
Thanks so much for working so hard on these, @lauramedalia, and thank you for your editing @Simek! This looks good from here. Let's get it into the site :)

rachelnabors commented:
PARTY!
